### PR TITLE
ADM-359

### PR DESCRIPTION
**Pull request by fengxinhou**
## Summary

Board Section

## Before
image

## After

image

### CODE REVIEW

File: frontend/__tests__/src/components/metrics/ConfigStep/Board.test.tsx [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    })
+
+    fireEvent.mouseDown(getByRole('button', { name: 'board' }))
+    fireEvent.click(getByText(BOARD_TYPES.LINEAR))
+
+    await waitFor(() => {
+      expect(getByText(BOARD_TYPES.LINEAR)).toBeInTheDocument()
+    })
+  })
+  it('should show error message when input a wrong type email ', async () => {
```
Comment (guzhongren) [inline frontend/__tests__/src/components/metrics/ConfigStep/Board.test.tsx:50]: `input a wrong type email` or `input a wrong email type`?

File: frontend/__tests__/src/components/metrics/ConfigStep/Board.test.tsx [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    const { getByRole, getByText } = render(<Board />)
+    const emailInput = getByRole('textbox', {
+      name: 'email',
+    })
+
+    fireEvent.change(emailInput, { target: { value: 'wrong email type' } })
+
+    await waitFor(() => {
+      expect(getByText('email is required')).toBeVisible()
```
Comment (guzhongren) [inline frontend/__tests__/src/components/metrics/ConfigStep/Board.test.tsx:59]: `email is required`, duplicated code

File: frontend/__tests__/src/components/metrics/ConfigStep/MetricsTypeCheckbox.test.tsx [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    fireEvent.mouseDown(getByRole('listbox', { name: REQUIRE_DATA }))
+
+    const errorMessage = getByText('Metrics is required')
+    expect(errorMessage).toBeInTheDocument()
+  })
+
+  it('should call handleRequireDate function when click MetricsTypeCheckbox selection, ', async () => {
+    const handleRequireDate = jest.fn()
+    const { getByRole } = render(<MetricsTypeCheckbox onHandleRequireData={handleRequireDate} />)
+    fireEvent.mouseDown(getByRole('button', { name: REQUIRE_DATA }))
```
Comment (guzhongren) [inline frontend/__tests__/src/components/metrics/ConfigStep/MetricsTypeCheckbox.test.tsx:50]: REQUIRED_DATA

File: frontend/__tests__/src/components/metrics/ConfigStep/MetricsTypeCheckbox.test.tsx [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    const listBox = within(getByRole('listbox'))
+    fireEvent.click(listBox.getByRole('option', { name: VELOCITY }))
+
+    expect(handleRequireDate).toBeCalled()
```
Comment (guzhongren) [inline frontend/__tests__/src/components/metrics/ConfigStep/MetricsTypeCheckbox.test.tsx:54]: Should confirm the function has been called `n` times, and the parameters

Comment (guzhongren) [inline frontend/__tests__/src/components/metrics/ConfigStep/MetricsTypeCheckbox.test.tsx:54]: Should confirm the function has been called `n` times, and the parameters

File: frontend/e2e/src/pages/configStep.test.ts [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+const COLLECTION_DATA = 'Collection Data'
+const FORM_LABEL = 'From *'
+const CHOOSE_DATA = 'Choose date'
+const TO_LABEL = 'To *'
+const ERROR_DATE = '02/20'
+const REQUIRED_DATA = 'Required Data'
+const VELOCITY = 'Velocity'
+const CLASSIFICATION = 'Classification'
+
+const clickCreateNewProjectButton = async (page) => {
```
Comment (guzhongren) [inline frontend/e2e/src/pages/configStep.test.ts:15]: PLZ move some methods like `select` and `fill fields` into a file to share with other steps or flow, you can learn PageObject Model

File: frontend/src/components/metrics/ConfigStep/Board/index.tsx [hunk 6]
```diff
@@ -0,0 +1,83 @@
+import { InputLabel, ListItemText, MenuItem, Select } from '@mui/material'
+import { BOARD_FIELDS, BOARD_TYPES, emailRegExp, ZERO } from '@src/constants'
```
Comment (guzhongren) [inline frontend/src/components/metrics/ConfigStep/Board/index.tsx:2]: `EMAIL_REG_EXP`

File: frontend/src/components/metrics/ConfigStep/Board/index.tsx [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+  BoardForm,
+  BoardSection,
+  BoardTextField,
+  BoardTitle,
+  BoardTypeSelections,
+  VerifyButton,
+} from '@src/components/metrics/ConfigStep/Board/style'
+
+export const Board = () => {
+  const [boardField, seBoardField] = useState({
```
Comment (guzhongren) [inline frontend/src/components/metrics/ConfigStep/Board/index.tsx:15]: Should store those fields into Redux, coc we will use those filed later, like export data

File: frontend/src/components/metrics/ConfigStep/Board/style.tsx [hunk 8]
```diff
@@ -0,0 +1,41 @@
+import { styled } from '@mui/material/styles'
+import { Button, FormControl, TextField } from '@mui/material'
+
+export const BoardSection = styled('div')({
+  boxShadow: '0 2px 1px -1px rgb(0 0 0 / 20%), 0 1px 1px 0 rgb(0 0 0 / 14%), 0 1px 3px 0 rgb(0 0 0 / 12%);',
```
Comment (guzhongren) [inline frontend/src/components/metrics/ConfigStep/Board/style.tsx:5]: Could we covert `px` to `rem`?

File: frontend/src/components/metrics/ConfigStep/DateRangePicker/index.tsx [hunk 9]
```diff
@@ -36,10 +35,10 @@ export const DateRangePicker = () => {
           PaperProps={datePickerPropsStyles}
           label='From'
           value={dateRange.startDate}
-          disablePast={true}
           onChange={(newValue) => {
             checkDateformat(newValue, DATE_RANGE.START_DATE)
-            checkDateRangeValid(newValue, dateRange.endDate)
+            checkDateRangeValid(newValue, newValue && newValue.add(14, 'day'))
+            setDateRange({ startDate: newValue, endDate: newValue && newValue.add(14, 'day') })
```
Comment (guzhongren) [inline frontend/src/components/metrics/ConfigStep/DateRangePicker/index.tsx:41]: Recommend move event handler into a function out of `jsx`

File: frontend/src/components/metrics/ConfigStep/MetricsTypeCheckbox/index.tsx [hunk 10]
```diff
@@ -0,0 +1,45 @@
+import { Checkbox, FormHelperText, InputLabel, ListItemText, MenuItem, Select, SelectChangeEvent } from '@mui/material'
+import { REQUIRE_DATA } from '@src/constants'
+import React, { useState } from 'react'
+import { RequireDataSelections } from '@src/components/metrics/ConfigStep/MetricsTypeCheckbox/style'
+interface MetricsTypeCheckboxProps {
+  onHandleRequireData: (value: string[]) => void
```
Comment (guzhongren) [inline frontend/src/components/metrics/ConfigStep/MetricsTypeCheckbox/index.tsx:6]: `onHandleRequiredData`

File: frontend/src/components/metrics/ConfigStep/index.tsx [hunk 11]
```diff
@@ -1,31 +1,21 @@
...
-  ListItemText,
-  MenuItem,
-  Radio,
-  RadioGroup,
-  Select,
-  SelectChangeEvent,
-} from '@mui/material'
+import { FormControlLabel, Radio, RadioGroup } from '@mui/material'
 import React, { useState } from 'react'
-import { CHINA_CALENDAR, REGULAR_CALENDAR, REQUIRE_DATA, STEPS } from '@src/constants'
```
Comment (guzhongren) [inline frontend/src/components/metrics/ConfigStep/index.tsx:14]: `REQUIRED_DATA`

File: frontend/src/components/metrics/ConfigStep/index.tsx [hunk 12]
```diff
@@ -1,31 +1,21 @@
...
+import { CHINA_CALENDAR, REGULAR_CALENDAR, STEPS } from '@src/constants'
 import { DateRangePicker } from '@src/components/metrics/ConfigStep/DateRangePicker'
-import { BackButton, ButtonGroup, ExportButton, NextButton, ProjectNameInput, RequireDataSelections } from './style'
+import { BackButton, ButtonGroup, ConfigStepWrapper, ExportButton, NextButton, ProjectNameInput } from './style'
 import { useAppDispatch, useAppSelector } from '@src/hooks'
 import { backStep, nextStep, selectStep } from '@src/features/stepper/StepperSlice'
+import { Board } from '@src/components/metrics/ConfigStep/Board'
+import { MetricsTypeCheckbox } from '@src/components/metrics/ConfigStep/MetricsTypeCheckbox'
+import { REQUIRE_DATAS } from '../../../../__tests__/src/fixtures'
```
Comment (guzhongren) [inline frontend/src/components/metrics/ConfigStep/index.tsx:10]: Should NOT use test code in PROD code!!!!

File: frontend/src/components/metrics/ConfigStep/index.tsx [hunk 13]
```diff
@@ -1,31 +1,21 @@
...
 
 export const ConfigStep = () => {
   const dispatch = useAppDispatch()
   const activeStep = useAppSelector(selectStep)
 
   const [projectName, setProjectName] = useState<string>('')
```
Comment (guzhongren) [inline frontend/src/components/metrics/ConfigStep/index.tsx:24]: Should store config info into Redux, not in component

File: frontend/src/components/metrics/ConfigStep/style.tsx [hunk 14]
```diff
@@ -1,7 +1,7 @@
 import { theme } from '@src/theme'
 import { styled } from '@mui/material/styles'
 import Button from '@mui/material/Button'
-import { FormControl, TextField } from '@mui/material'
+import { TextField } from '@mui/material'
 
 const basicButtonStyle = {
   boxShadow: '0 3px 1px -2px rgb(0 0 0 / 20%), 0 2px 2px 0 rgb(0 0 0 / 14%), 0 1px 5px 0 rgb(0 0 0 / 12%)',
```
Comment (guzhongren) [inline frontend/src/components/metrics/ConfigStep/style.tsx:7]: `px` to `rem`

File: frontend/src/components/metrics/MetricsStepper/style.tsx [hunk 15]
```diff
@@ -9,6 +9,7 @@ export const MetricsStepLabel = styled(StepLabel)({
 export const MetricsStepperContent = styled('div')({
   display: 'flex',
   flexDirection: 'column',
-  margin: '0 280px',
```
Comment (guzhongren) [inline frontend/src/components/metrics/MetricsStepper/style.tsx:12]: `px` to `rem`

**Review by guzhongren — CHANGES_REQUESTED** [review]
PLZ update with the comments

File: frontend/__tests__/src/components/metrics/ConfigStep/Board.test.tsx (hunk 1 above)
Comment (fengxinhou) [reply frontend/__tests__/src/components/metrics/ConfigStep/Board.test.tsx:50]: wrong  type email

**Comment by guzhongren** [comment]
fix lint warnings

**Review by guzhongren — CHANGES_REQUESTED** [review]
- fix lint warnings
- remove mock server and config
- should not change backend code

File: frontend/__tests__/src/components/metrics/ConfigStep/Board.test.tsx [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+    fireEvent.click(getByText('Verify'))
+
+    expect(getByText('Reset')).toBeVisible()
+  })
+  it('should called verifyBoard method once when click verify button', async () => {
+    const { getByRole } = setup()
+    fillBoardFieldsInformation()
+    fireEvent.click(getByRole('button', { name: 'Verify' }))
+
+    expect(() => expect(verifyBoard).toHaveBeenCalledTimes(1)).toThrow()
```
Comment (guzhongren) [inline frontend/__tests__/src/components/metrics/ConfigStep/Board.test.tsx:152]: Why it throw excepiton?

File: frontend/src/service/config.service.ts [hunk 17]
```diff
@@ -0,0 +1,4 @@
+import axios from 'axios'
+
+const URL = 'http://localhost:3000/verifyBoard'
```
Comment (guzhongren) [inline frontend/src/service/config.service.ts:3]: The URL is a good pattern? PLZ learn Restful API best practice

File: frontend/__tests__/src/components/metrics/ConfigStep/Board.test.tsx (hunk 16 above)
Comment (fengxinhou) [reply frontend/__tests__/src/components/metrics/ConfigStep/Board.test.tsx]: replace by  expect(verifyBoard).toHaveBeenCalledTimes(1)

Comment (fengxinhou) [reply frontend/__tests__/src/components/metrics/ConfigStep/Board.test.tsx]: Becase the logic code only calls the method and dose not process the obtained data, throw is used ensure that the test passes. However, it turns out that the method can be directly tested once. Now, it has been modified.

File: frontend/src/service/config.service.ts (hunk 17 above)
Comment (fengxinhou) [reply frontend/src/service/config.service.ts]: delete json-server and replace with a fake road address https://jsonplaceholder.typicode.com/posts to impersonation request

File: frontend/__tests__/src/components/metrics/ConfigStep/MetricsTypeCheckbox.test.tsx [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+    const listBox = within(getByRole('listbox'))
+    fireEvent.click(listBox.getByRole('option', { name: VELOCITY }))
+    fireEvent.click(listBox.getByRole('option', { name: VELOCITY }))
+    fireEvent.mouseDown(getByRole('listbox', { name: REQUIRED_DATA }))
+
+    const errorMessage = getByText('Metrics is required')
+    expect(errorMessage).toBeInTheDocument()
+  })
+
+  it('should board component when click MetricsTypeCheckbox selection velocity ', () => {
```
Comment (mengna-tw) [inline frontend/__tests__/src/components/metrics/ConfigStep/MetricsTypeCheckbox.test.tsx:61]: maybe update description as should show board component....

File: frontend/__tests__/src/features/boardSlice.test.tsx [hunk 19]
```diff
@@ -0,0 +1,20 @@
+import boardReducer, { changeBoardVerifyState } from '@src/features/board/boardSlice'
+
+describe('board reducer', () => {
+  it('should false when handle initial state', () => {
+    const STEPPER = boardReducer(undefined, { type: 'unknown' })
```
Comment (mengna-tw) [inline frontend/__tests__/src/features/boardSlice.test.tsx:5]: maybe the description update to 'should return false', and the constant update as stepper

File: frontend/__tests__/src/features/boardSlice.test.tsx [hunk 20]
```diff
@@ -0,0 +1,20 @@
...
+
+    expect(STEPPER.isBoardVerified).toEqual(false)
+  })
+
+  it('should true when handle changeBoardVerifyState given isBoardVerified is true', () => {
```
Comment (mengna-tw) [inline frontend/__tests__/src/features/boardSlice.test.tsx:10]: should return true....

File: frontend/__tests__/src/features/configSlice.test.tsx [hunk 21]
```diff
@@ -0,0 +1,131 @@
...
+        boardFields: {
+          board: BOARD_TYPES.JIRA,
+          boardId: '',
+          email: '',
+          projectKey: '',
+          site: '',
+          token: '',
+        },
+      },
+
```
Comment (mengna-tw) [inline frontend/__tests__/src/features/configSlice.test.tsx:36]: should remove this empty line

**Review by guzhongren — APPROVED** [review]

File: frontend/__tests__/src/services/BoardService.test.tsx [hunk 22]
```diff
@@ -0,0 +1,40 @@
...
+      ctx.json({
+        errorMessage: 'error message',
+      })
+    )
+  })
+)
+
+describe('board service successfully', () => {
+  beforeAll(() => server.listen())
+  afterAll(() => server.close())
```
Comment (mengna-tw) [inline frontend/__tests__/src/services/BoardService.test.tsx:26]: maybe move line 25,26 out of describe, and update fn to beforeEach and afterEach

**Review by mengna-tw — COMMENTED** [review]
left some small comments, feel free to fix them

File: frontend/src/store/store.ts [hunk 23]
```diff
@@ -1,8 +1,12 @@
 import { configureStore } from '@reduxjs/toolkit'
 import stepperReducer from '../features/stepper/StepperSlice'
+import configReducer from '../features/config/configSlice'
```
[the remark this answers is not in the document]
Comment (fengxinhou) [reply frontend/src/store/store.ts]: already changed ：move the store file to the root directory

File: frontend/__tests__/src/services/BoardService.test.tsx (hunk 22 above)
Comment (fengxinhou) [reply frontend/__tests__/src/services/BoardService.test.tsx:26]: https://mswjs.io/docs/api/setup-server  Refer to msw official website

File: frontend/__tests__/src/components/metrics/ConfigStep/Board.test.tsx [hunk 24]
```diff
@@ -1,10 +1,16 @@
...
-import { setupServer, SetupServerApi } from 'msw/node'
 import { Board } from '@src/components/metrics/ConfigStep/Board'
 import { BOARD_FIELDS, BOARD_TYPES, CONFIG_TITLE, ERROR_MESSAGE_COLOR } from '../../../fixtures'
 import { Provider } from 'react-redux'
 import { setupStore } from '../../../utils/setupStoreUtil'
+import { setupServer } from 'msw/node'
+import { rest } from 'msw'
+
+const server = setupServer(
+  rest.get('https://jsonplaceholder.typicode.com/posts', (req, res, ctx) => {
```
[the remark this answers is not in the document]
Comment (LanxinWang) [reply frontend/__tests__/src/components/metrics/ConfigStep/Board.test.tsx]: alredy done

[the remark this answers is not in the document]
Comment (LanxinWang) [reply frontend/__tests__/src/components/metrics/ConfigStep/Board.test.tsx]: already done

File: frontend/__tests__/src/service/BoardService.test.tsx [hunk 25]
```diff
@@ -0,0 +1,17 @@
+import { verifyBoard } from '@src/service/boardService'
+import { setupServer } from 'msw/node'
+import { rest } from 'msw'
+
+const server = setupServer(
+  rest.get('https://jsonplaceholder.typicode.com/posts', (req, res, ctx) => {
```
[the remark this answers is not in the document]
Comment (fengxinhou) [reply frontend/__tests__/src/service/BoardService.test.tsx]: alreday done

File: frontend/__tests__/src/components/metrics/ConfigStep/ConfigStep.test.tsx [hunk 26]
```diff
@@ -106,19 +98,17 @@ describe('ConfigStep', () => {
...
-
   it('should show board component when MetricsTypeCheckbox select  Classification, ', () => {
     const { getByRole } = setup()
 
     fireEvent.mouseDown(getByRole('button', { name: REQUIRED_DATA }))
     const requireDateSelection = within(getByRole('listbox'))
     fireEvent.click(requireDateSelection.getByRole('option', { name: 'Classification' }))
 
-    expect(getByRole('heading', { name: CONFIG_TITLE.BOARD, hidden: true })).toBeInTheDocument()
+    expect(getByRole('heading', { name: 'board', hidden: true })).toBeInTheDocument()
```
[the remark this answers is not in the document]
Comment (LanxinWang) [reply frontend/__tests__/src/components/metrics/ConfigStep/ConfigStep.test.tsx]: already done

**Pull request merged**